processing/src/main/java/org/apache/druid/segment/data/FrontCodedIndexed.java

```diff
@@ -0,0 +1,491 @@
+    final boolean hasNull = NullHandling.IS_NULL_BYTE == orderedBuffer.get();
+    final int numValues = VByte.readInt(orderedBuffer);
+    // size of offsets + values
+    final int size = VByte.readInt(orderedBuffer);
+    final int offsetsPosition = orderedBuffer.position();
+    // move position to end of buffer
+    buffer.position(offsetsPosition + size);
+
+    final int numBuckets = (int) Math.ceil((double) numValues / (double) bucketSize);
+    final int adjustIndex = hasNull ? 1 : 0;
+    final int div = Integer.numberOfTrailingZeros(bucketSize);
+    final int rem = bucketSize - 1;
+    return () -> new FrontCodedIndexed(
+        orderedBuffer,
+        bucketSize,
+        numBuckets,
+        (numValues & rem) == 0 ? bucketSize : numValues & rem,
+        hasNull,
+        numValues + adjustIndex,
+        adjustIndex,
+        div,
+        rem,
+        offsetsPosition,
+        offsetsPosition + ((numBuckets - 1) * Integer.BYTES)
+    );
+  }
+
+  private final ByteBuffer buffer;
+  private final int adjustedNumValues;
+  private final int adjustIndex;
+  private final int bucketSize;
+  private final int numBuckets;
+  private final int div;
+  private final int rem;
+  private final int offsetsPosition;
+  private final int bucketsPosition;
+  private final boolean hasNull;
+  private final int lastBucketNumValues;
+
+  private FrontCodedIndexed(
+      ByteBuffer buffer,
+      int bucketSize,
+      int numBuckets,
+      int lastBucketNumValues,
+      boolean hasNull,
+      int adjustedNumValues,
+      int adjustIndex,
+      int div,
+      int rem,
+      int offsetsPosition,
+      int bucketsPosition
+  )
+  {
+    if (Integer.bitCount(bucketSize) != 1) {
+      throw new ISE("bucketSize must be a power of two but was[%,d]", bucketSize);
+    }
+    this.buffer = buffer.asReadOnlyBuffer().order(buffer.order());
+    this.bucketSize = bucketSize;
+    this.hasNull = hasNull;
+    this.adjustedNumValues = adjustedNumValues;
+    this.adjustIndex = adjustIndex;
+    this.div = div;
+    this.rem = rem;
+    this.numBuckets = numBuckets;
+    this.offsetsPosition = offsetsPosition;
+    this.bucketsPosition = bucketsPosition;
+    this.lastBucketNumValues = lastBucketNumValues;
+  }
+
+  @Override
+  public int size()
+  {
+    return adjustedNumValues;
+  }
+
+  @Nullable
+  @Override
+  public ByteBuffer get(int index)
+  {
+    if (hasNull && index == 0) {
+      return null;
+    }
+
+    // due to vbyte encoding, the null value is not actually stored in the bucket (no negative values), so we adjust
+    // the index
+    final int adjustedIndex = index - adjustIndex;
+    // find the bucket which contains the value with maths
+    final int bucket = adjustedIndex >> div;
+    final int bucketIndex = adjustedIndex & rem;
+    final int offset = getBucketOffset(bucket);
+    buffer.position(offset);
+    return getFromBucket(buffer, bucketIndex);
+  }
+
+  @Override
+  public int indexOf(@Nullable ByteBuffer value)
```

Comment:
Javadoc should explicitly tighten the contract such that this returns `(-(insertion point) - 1)` on no match.

Reply:
yeah, i noticed this too. Part of the problem is that classically this isn't strictly required in all uses of  `Indexed.indexOf` but looking around there is only 1 usage/implementation which couldn't implement this contract, in https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/segment/DictionaryEncodedColumnMerger.java#L148 which is using it to determine if the dictionary contains a null value when merging, which when backed by the string indexer is getting a -1 for any value that isn't in the dictionary, https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/segment/DictionaryEncodedColumnIndexer.java#L96.

This could usage probably be replaced with a method to indicate if the value dictionary contains null and we could strengthen the contract of indexOf since all remaining users would be related to column value dictionary stuff that requires it. I wonder if we should also consider adding a method to check that the `Indexed` is sorted, since that `indexOf` contract only really makes sense for finding endpoints for value ranges using the same comparator as the dictionary is sorted.

I considered making a `ValueDictionary` interface with a stronger contract around this behavior, but @cheddar pushed back asserting we already have too many interfaces which .. is fair.

Thoughts on pushing the extra methods into `Indexed` so that the contract can be strengthened to `(-(insertion point) - 1)` if `isSorted()` returns true or something and `hasNull` to allow the string merger to do the check it needs?

Reply:
I ended up adding `isSorted` which when true strengthens the contract of `indexOf` to be `(-(insertion point) - 1)`. `GenericIndexed` (and `BufferedIndexed` produced by its `singleThreaded`), `FixedIndexed`, and `FrontCodedIndexed` all implement it.